### ISSUE TITLE
Unit testing: calling `APIClient.force_authenticate()` with `user=None` does not work as intended

### ISSUE THREAD

**Issue by willbeaufoy** [encode/django-rest-framework#8211]
### Discussed in https://github.com/encode/django-rest-framework/discussions/8184

<div type='discussions-op-text'>

<sup>Originally posted by **willbeaufoy** September 24, 2021</sup>
If you call [`APIClient.force_authenticate()`](https://github.com/encode/django-rest-framework/blob/master/rest_framework/test.py#L273) with a `token` param but without a `user` param, `self.handler._force_token` is set to the provided token, but then `self.logout()` is [called](https://github.com/encode/django-rest-framework/blob/master/rest_framework/test.py#L281), which [immediately](https://github.com/encode/django-rest-framework/blob/master/rest_framework/test.py#L339) sets `self.handler._force_token` to `None` again. Surely this is not intended? The docstring and the [docs](https://www.django-rest-framework.org/api-guide/testing/#forcing-authentication) say you can use either a user or a token or both, but in reality you cannot just use a token.

I discovered this while writing a unit test for an endpoint with access authorised by the OAuth client credentials method (provided by django-oauth-toolkit), where requests have a token but no user. Therefore I tried to authenticate then call my endpoint like this:

```python
from rest_framework.test import APITestCase

class MyTests(APITestCase):
    def test_endpoint(self):
        # ... Code to create valid access_token here
        self.client.force_authenticate(token=access_token)

        response = self.client.get("/my/endpoint/")

        self.assertEqual(response.status_code, HTTPStatus.OK)
```

But my test fails as I get a 401 Unauthorized response.

However if I comment out the [line](https://github.com/encode/django-rest-framework/blob/master/rest_framework/test.py#L339) in `APIClient.logout()` that sets `self.handler._force_token` to `None`, the request makes it through to my endpoint successfully and my test passes.

The [PR](https://github.com/encode/django-rest-framework/pull/2259) that changed the `logout()` method to set `self.handler._force_user` and `self.handler._force_token` to `None` was done for an unrelated reason. Perhaps at the time it was overlooked that this broke the case I am trying to test above?

If you agree that this current apparently broken behaviour is a bug, then I can do a PR to fix it.</div>

**Comment by willbeaufoy** [comment]
Created an issue from this as I'm doing a PR to fix it.

**Comment by stcalica** [comment]
having this issue as well -- any updates?

**Comment by willbeaufoy** [comment]
> having this issue as well -- any updates?

Thanks for the prompt, I just addressed the outstanding comment so hopefully it can be merged soon.

**Comment by tomchristie** [comment]
Closed via #8212
